### PR TITLE
Migration to NodeJS

### PR DESCRIPTION
**Pull request by SaiArrow**

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -5,20 +5,16 @@ A demo of how to run Vega with a PostgreSQL backend. This is a fork of [this pro
 ## Installation
 1. Install and start PostgreSQL.
 2. Create a PostgreSQL database named `scalable_vega`, e.g., `createdb scalable_vega`
+    1. You don't need to install or setup for DuckDB
```
Comment (domoritz) [inline README.md:8]:
```suggestion
2. Create a PostgreSQL database named `scalable_vega`, e.g., `createdb scalable_vega`. You don't need to do anything if you use DuckDB.
```

File: README.md [hunk 2]
```diff
@@ -5,20 +5,16 @@ A demo of how to run Vega with a PostgreSQL backend. This is a fork of [this pro
...
 3. Run `cd /path/to/dev/repos`.
 4. Run `git clone git@github.com:leibatt/scalable-vega.git`.
-5. Run `pip install requirements.txt` to install python dependencies. (Python 3.x is preferable)
-6. Run `yarn` to javacript install dependencies.
-7. Install Jest (for running Tests): `yarn add --dev jest`. <br>
-8. For using prepopulated database, look at additional notes.
+5. Run `yarn` to javacript install dependencies.
```
Comment (domoritz) [inline README.md:11]:
```suggestion
5. Run `yarn` to install NPM dependencies.
```

File: README.md [hunk 3]
```diff
@@ -5,20 +5,16 @@ A demo of how to run Vega with a PostgreSQL backend. This is a fork of [this pro
...
 
 ## Running Application Server
-1. Run `python server.py --CI --db [postgresql/duckdb]` <br>
-&nbsp;&nbsp;&nbsp; a. `--CI` refers to the setup for github actions. <br>
-&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; i. Also if you want to connect to postgres using a password (peer/md5/scram-sha-256 connection), use this flag. <br>
-&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; ii. If you don't want to use a password (trust connection), don't use this flag. <br>
-&nbsp;&nbsp;&nbsp; b. `--db` refers to the DBMS currently being used, default is PostgreSQL. For DuckDB, use `--db duckdb`<br>
-2. All the config information for the databases and server (including user, password and ports to be used) is currently stored in the `config` folder and can be customized.
+1. Run `node server.js` 
+    1. Default database used is DuckDB, if you want to use postgres as your database run `node server.js pg`
```
Comment (domoritz) [inline README.md:16]:
```suggestion
    1. The default database used is DuckDB. If you want to use postgres as your database run `node server.js pg`.
```

File: lib/spec_rewrite.ts [hunk 4]
```diff
@@ -1,6 +1,12 @@
 import * as Vega from "vega"
 import { Transforms, AggregateTransform, None } from "vega"
 import { aggregate, extent, bin } from "vega-transforms";
+import { parse } from "vega-expression"
+//import expr2sql from "./expr2sql"
```
Comment (domoritz) [inline lib/spec_rewrite.ts:5]:
```suggestion
```

File: lib/spec_rewrite.ts [hunk 5]
```diff
@@ -70,10 +76,11 @@ function vegaNonTransformToSql(tableName: string, markFields: string[]) {
   return `'${out}'`;
 }
 
-export const aggregateTransformToSql = (tableName: string, transform: AggregateTransform, db: string) => {
+export const aggregateTransformToSql = (tableName: string, transform: any, db: string, prev: any) => {
```
Comment (domoritz) [inline lib/spec_rewrite.ts:79]: Why any?

File: lib/spec_rewrite.ts [hunk 6]
```diff
@@ -116,7 +123,7 @@ function collectNonTransformFields(dataName: string, marks: any) {
   return fields
 }
 
-export function dataRewrite(tableName: string, transform: Transforms, db: string, dbTransforms, newData) {
+export function dataRewrite(tableName: string, transform: any, db: string, dbTransforms, newData) {
```
Comment (domoritz) [inline lib/spec_rewrite.ts:126]: Why any?

File: lib/spec_rewrite.ts [hunk 7]
```diff
@@ -169,26 +176,105 @@ export function dataRewrite(tableName: string, transform: Transforms, db: string
...
 
 }
+const filterTransformToSql = (tableName: string, transform: any, db: string, prev: any) => {
+  console.log(parse(transform.expr))
+  const filter = expr2sql(parse(transform.expr))
+  tableName = prev ? `(${prev.query.signal.slice(1, -1)}) ${prev.name}` : tableName
+
+
+  var sql = ''
+  sql = [
```
Comment (domoritz) [inline lib/spec_rewrite.ts:214]:
```suggestion
  const sql = [
```

File: lib/spec_rewrite.ts [hunk 8]
```diff
@@ -169,26 +176,105 @@ export function dataRewrite(tableName: string, transform: Transforms, db: string
...
+      return visit(n.left) + ' ' + n.operator + ' ' + visit(n.right)
+    },
+  }
+  return visit(expr)
+}
 
 export function specRewrite(vgSpec) {
   const dataSpec = vgSpec.data
-  const dbTransformInd = []
+  const dbTransformInd = []   //the data item to be removed
```
Comment (domoritz) [inline lib/spec_rewrite.ts:273]: from what?

File: lib/spec_rewrite_copy.ts [hunk 9]
```diff
@@ -0,0 +1,348 @@
+import * as Vega from "vega"
```
Comment (domoritz) [inline lib/spec_rewrite_copy.ts:1]: Why this copy?

File: server.js [hunk 10]
```diff
@@ -0,0 +1,279 @@
+"use strict";
```
Comment (domoritz) [inline server.js:1]: Why is this compiled file in the repo?

File: server.ts [hunk 11]
```diff
@@ -0,0 +1,236 @@
...
+    console.log("Hello", results['rows']);
+    res.status(200).send(results['rows']);
+    }
+    else {
+    client = await db.connect();
+    await client.all(query, function(err, results) {
+  	if (err) {
+    		throw err;
+  	}
+  	console.log("DuckDb", results)
```
Comment (domoritz) [inline server.ts:63]: mix tab and spaces

File: server.ts [hunk 12]
```diff
@@ -0,0 +1,236 @@
...
+  // Client needs to send more type data in this case.
+  const type = typeof value;
+  if (type === 'string') {
+    return 'VARCHAR(256)';
+  } else if (type === 'number') {
+    return 'FLOAT';
+  } else if (type === 'boolean') {
+    return 'BOOLEAN';
+  } else {
+    throw 'undefined type: \'' + type + '\'';
```
Comment (domoritz) [inline server.ts:92]: Use template literals

File: server.ts [hunk 13]
```diff
@@ -0,0 +1,236 @@
...
+    if (!schema.hasOwnProperty(attrName)) {
+      continue;
+    }
+    let attrType: string = schema[attrName];
+    if (first) {
+      first = false;
+    } else {
+      out += ', ';
+    }
+    out += (attrName + ' ' + attrType)
```
Comment (domoritz) [inline server.ts:119]: Why the ()?

File: test/filter_pg.test.ts [hunk 14]
```diff
@@ -0,0 +1,151 @@
+import { specRewrite } from "../lib/spec_rewrite"
+//import { VegaDbTransform } from "../lib/dbtransform"
+//var vega = require('vega')
```
Comment (domoritz) [inline test/filter_pg.test.ts:3]: rm

File: test/filter_pg.test.ts [hunk 15]
```diff
@@ -0,0 +1,151 @@
...
+    var loader = vega.loader();
+
+    var view = new vega.View(vega.parse(spec_vg), {
+      loader: loader,
+      renderer: 'none'
+    });
+    await view.runAsync();
+
+    var result_vg = view.data('cars');
+    // console.log(result_vg);
```
Comment (domoritz) [inline test/filter_pg.test.ts:123]: rm comments

File: test/filter_pg.test.ts [hunk 16]
```diff
@@ -0,0 +1,151 @@
...
+
+    var view_s = new vega.View(runtime, {
+      renderer: 'none'
+    })
+      .logLevel(vega.Info)
+
+    await view_s.runAsync();
+
+    var result_s = view_s.data('cars');
+    // console.log(result_s);
```
Comment (domoritz) [inline test/filter_pg.test.ts:146]: same

File: server.js [hunk 17]
```diff
@@ -0,0 +1,279 @@
...
+                        client.release();
+                    }
+                }
+                console.log("Final");
+                return [7 /*endfinally*/];
+            case 9: return [2 /*return*/];
+        }
+    });
+}); });
+function postgresTypeFor(value) {
```
Comment (domoritz) [inline server.js:125]: duplicated code

**Review by domoritz — CHANGES_REQUESTED** [review]

File: lib/spec_rewrite.ts (hunk 5 above)
Comment (SaiArrow) [reply lib/spec_rewrite.ts:78]: @JunranY, could you take a look at this ?

**Comment by domoritz** [comment]
There are a lot of conflicts. Please make sure to merge the current master branch.

File: main.ts [hunk 18]
```diff
@@ -0,0 +1,122 @@
...
+
+const querystring = require('querystring');
+const http = require('http');
+
+
+export function run(spec: vega.Spec) {
+  // (re-)run vega using the scalable vega version
+  // FixMe: should we define these attributes in the spec somehow?
+  const httpOptions = {
+    "url": 'http://localhost:3000/query',
```
Comment (domoritz) [inline main.ts:16]: Hardcoding localhost is not good long term.

File: package.json [hunk 19]
```diff
@@ -21,7 +21,8 @@
     "vega-encode": "^4.8.0",
     "vega-transform-db": "^1.0.0",
     "vega-transform-omnisci-core": "^0.0.10",
-    "vega-transforms": "^4.9.0"
+    "vega-transforms": "^4.9.0",
```
Comment (domoritz) [inline package.json:24]: Use a yarn.lock, mot package.lock file above.

**Review by domoritz — CHANGES_REQUESTED** [review]

File: lib/spec_rewrite.ts (hunk 5 above)
Comment (domoritz) [reply lib/spec_rewrite.ts]: This is a change in this pull request so I think you could revert it and see what happens.

Comment (JunranY) [reply lib/spec_rewrite.ts]: I reverted it through my last pr https://github.com/leibatt/scalable-vega/pull/20 (made before this one, but committed later)

Comment (domoritz) [reply lib/spec_rewrite.ts]: @SaiArrow then you can just merge master into this

**Comment by domoritz** [comment]
Should you delete the requirements.txt file as well?

**Comment by domoritz** [comment]
And do we still need the setup python in the test workflow?

**Comment by SaiArrow** [comment]
I wasn't sure if we plan on completing switching over to NodsJS right now, should I instead store all files related to python in a separate folder ?

**Comment by domoritz** [comment]
Are we still using Python? I thought this pull request migrates but maybe misunderstood it.

**Pull request merged**